services/device/geolocation/geoclue_client_object.cc

```diff
@@ -0,0 +1,235 @@
+                            base::BindOnce(&GeoClueClientObject::OnGotClient,
+                                           weak_ptr_factory_.GetWeakPtr()));
+}
+
+void GeoClueClientObject::OnGotClient(dbus::Response* response) {
+  if (!response) {
+    NotifyError("Failed to get a GeoClue2.Client object");
+    return;
+  }
+
+  dbus::MessageReader reader(response);
+  dbus::ObjectPath client_path;
+  if (!reader.PopObjectPath(&client_path)) {
+    NotifyError("Failed to read object path for GeoClue2.Client");
+    return;
+  }
+
+  proxy_ = creation_params_.bus->GetObjectProxy(kServiceName,
+                                                dbus::ObjectPath(client_path));
+  SetProperties();
+}
+
+void GeoClueClientObject::SetProperties() {
+  CHECK(proxy_);
+
+  properties_ = std::make_unique<Properties>(proxy_.get());
+
+  const auto callback = base::BarrierCallback<bool>(
+      2, base::BindOnce(&GeoClueClientObject::OnSetProperties,
+                        weak_ptr_factory_.GetWeakPtr()));
+
+  properties_->requested_accuracy_level.Set(creation_params_.high_accuracy
+                                                ? AccuracyLevel::kExact
+                                                : AccuracyLevel::kCity,
+                                            callback);
+  properties_->desktop_id.Set(creation_params_.desktop_id, callback);
+}
+
+void GeoClueClientObject::OnSetProperties(std::vector<bool> results) {
+  CHECK(properties_);
+
+  if (base::ranges::any_of(results, [](const auto& value) { return !value; })) {
+    NotifyError("Failed to set desktop_id and accuracy level.");
+    return;
+  }
+
+  ConnectLocationUpdated();
+}
+
+void GeoClueClientObject::ConnectLocationUpdated() {
+  CHECK(proxy_);
+
+  proxy_->ConnectToSignal(
+      kInterfaceName, "LocationUpdated",
+      base::BindRepeating(&GeoClueClientObject::OnLocationUpdated,
+                          weak_ptr_factory_.GetWeakPtr()),
+      base::BindOnce(&GeoClueClientObject::OnLocationUpdatedConnected,
+                     weak_ptr_factory_.GetWeakPtr()));
+}
+
+void GeoClueClientObject::OnLocationUpdatedConnected(
+    const std::string& service,
+    const std::string& interface,
+    bool success) {
+  if (!success) {
+    NotifyError("Failed to connect to LocationUpdate signal");
+    return;
+  }
+
+  state_ = kInitialized;
+  MaybeStartClient();
+}
+
+void GeoClueClientObject::OnLocationUpdated(dbus::Signal* signal) {
+  dbus::MessageReader reader(signal);
+  dbus::ObjectPath old_location;
+  dbus::ObjectPath new_location;
+  if (!reader.PopObjectPath(&old_location) ||
+      !reader.PopObjectPath(&new_location)) {
+    NotifyError("Failed to read location update");
+    return;
+  }
+  auto properties = std::make_unique<LocationProperties>(
+      creation_params_.bus->GetObjectProxy(kServiceName, new_location));
+  properties->GetAll(base::BindOnce(creation_params_.on_location_changed,
+                                    std::move(properties)));
```

Comment:
eh. you're moving and accessing the same instance. this is bad.

Reply:
:+1: my bad.

So I've added the `LocationProperties` as a class member. Originally I wasn't doing this because each LocationUpdate results in a new `LocationProperties` object at a new `ObjectPath`

However, I think this introduces a race:

1. `LocationUpdate` fires. Reset member to new `LocationProperties` at `/Location/1` and start reading values
2. `LocationUpdate` fires. Reset member to new `LocationProperties` at `/Location/2` and start reading values
3. Read of `/Location/1` completes, and we trigger the callback, but the `LocationProperties` object we have as a class member is a partially read `/Location/2`
4. Read of `/Location/2` completes and we trigger the callback with the fully read `/Location/2`

Would changing this back, but introducing a separate `properties_raw` pointer be okay?


```cpp
  auto properties = std::make_unique<LocationProperties>(
      creation_params_.bus->GetObjectProxy(kServiceName, new_location));
  auto* properties_raw = properties.get();
  properties_raw->GetAll(base::BindOnce(creation_params_.on_location_changed,
                                    std::move(properties)));
```

Reply:
> 2\. `LocationUpdate` fires. Reset member to new `LocationProperties` at `/Location/2` and start reading values

the first `LocationProperties` at this point will be destroyed, so there's no "read completes" event for `/Location/1`.

Technically the queued read on `dbus` thread will be performed, but it will be replied to a callback (1) which is weak_ptr-invalidated, so there's no receiving side.

1. https://github.com/chromium/chromium/blob/aa227489f353c34cc8348ff3958cc85db421a273/dbus/property.cc#L169

Reply:
Ah right, I was forgetting that when the DBus properties object is destroyed it won't fire any callbacks! Thanks @goodov